### PR TITLE
primer PR

### PR DESCRIPTION
**Pull request by kevinschweitzer**

### CODE REVIEW

File: app/src/main/java/com/example/kevinschweitzer/myapplication/mvp/model/CalculatorModel.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+public class CalculatorModel {
+
+    private String number1;
+    private String number2;
+    private Operator operator;
+
+    public CalculatorModel() {
+        number1 = "";
+        number2 = "";
```
Comment (nacho-globant) [inline app/src/main/java/com/example/kevinschweitzer/myapplication/mvp/model/CalculatorModel.java]: Avoid this initialization, and use the null initialization for these fields

File: app/src/main/java/com/example/kevinschweitzer/myapplication/mvp/model/CalculatorModel.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+            number2+=str;
+        }
+    }
+
+    public void setOperator(Operator operator){
+        this.operator = operator;
+    }
+
+    public Integer getResult(){
+        if(number1 != "" && number2 != ""){
```
Comment (nacho-globant) [inline app/src/main/java/com/example/kevinschweitzer/myapplication/mvp/model/CalculatorModel.java]: Replace number1 != "" for !TextUtils.isEmpty(number1)

The same for number2

File: app/src/main/java/com/example/kevinschweitzer/myapplication/mvp/presenter/CalculatorPresenter.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+     * Resolve the calculation
+     * If the equals symbol is clicked, this action must be done
+     */
+    public void onEqualsClicked(){
+        Integer result = model.getResult();
+        Log.d("RESULT",result+"");
+        if(result!=null){
+            view.setResult(result);
+        }else{
+            view.setResult(view.getActivity().getResources().getString(R.string.error));
```
Comment (nacho-globant) [inline app/src/main/java/com/example/kevinschweitzer/myapplication/mvp/presenter/CalculatorPresenter.java]: Avoid NPE for view.getActivity()

File: app/src/main/java/com/example/kevinschweitzer/myapplication/mvp/view/CalculatorView.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+import butterknife.BindView;
+import butterknife.ButterKnife;
+import butterknife.OnClick;
+
+public class CalculatorView {
+
+    @BindView(R.id.text_operation) TextView operation;
+    @BindView(R.id.text_result) TextView result;
+    private Activity activity;
```
Comment (nacho-globant) [inline app/src/main/java/com/example/kevinschweitzer/myapplication/mvp/view/CalculatorView.java]: Use WeakRefernece for this field, or use ActivityView directly

File: app/src/main/java/com/example/kevinschweitzer/myapplication/mvp/view/CalculatorView.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+    public CalculatorView(Activity activity){
+        this.activity = activity;
+        ButterKnife.bind(this, activity);
+    }
+
+    public void concatNumber(String str){
```
Comment (nacho-globant) [inline app/src/main/java/com/example/kevinschweitzer/myapplication/mvp/view/CalculatorView.java]: move this method to model

File: app/src/main/java/com/example/kevinschweitzer/myapplication/mvp/view/CalculatorView.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        operation.setText(operation.getText() + str);
+    }
+
+    public void concatOperator(String str){
```
Comment (nacho-globant) [inline app/src/main/java/com/example/kevinschweitzer/myapplication/mvp/view/CalculatorView.java]: move this method to model

File: app/src/main/java/com/example/kevinschweitzer/myapplication/mvp/view/CalculatorView.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    public void setResult(String error){
+        this.result.setText(error);
+    }
+
+    public void clear(){
+        operation.setText("");
+        result.setText("");
+    }
+
+    @OnClick({R.id.button_number0, R.id.button_number1, R.id.button_number2, R.id.button_number3, R.id.button_number4, R.id.button_number5,
```
Comment (nacho-globant) [inline app/src/main/java/com/example/kevinschweitzer/myapplication/mvp/view/CalculatorView.java:50]: This is fine, but you have another manner to this do the same in a more simple manner. You could move this to the activity and call the presenter.
You dont need change, its only a comment

File: app/src/main/res/layout/activity_main.xml [hunk 8]
```diff
@@ -0,0 +1,45 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (nacho-globant) [inline app/src/main/res/layout/activity_main.xml:1]: Use dimens.xml and strings.xml. Avoid hardocoded

Comment (nacho-globant) [inline app/src/main/res/layout/activity_main.xml:1]: Use dimens.xml and strings.xml. Avoid hardocoded values

File: app/src/main/res/layout/partial_numeric_panel.xml [hunk 9]
```diff
@@ -0,0 +1,150 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (nacho-globant) [inline app/src/main/res/layout/partial_numeric_panel.xml:1]: Use dimens.xml and strings.xml. Avoid hardocoded values

File: app/src/main/java/com/example/kevinschweitzer/myapplication/mvp/presenter/CalculatorPresenter.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+        model.setOperator(operator);
+        view.concatOperator(operator.getSymbol());
+    }
+
+    /**
+     * Resolve the calculation
+     * If the equals symbol is clicked, this action must be done
+     */
+    public void onEqualsClicked(){
+        Integer result = model.getResult();
```
Comment (nacho-globant) [inline app/src/main/java/com/example/kevinschweitzer/myapplication/mvp/presenter/CalculatorPresenter.java]: simplify this logic. For example

Activity activity = view.getActivity();
            if(activty==null){
                return;
        }
       Integer result = model.getResult();
            view.setResult(result!=null?result:activity.getResources().getString(R.string.error));

Comment (kevinschweitzer) [reply app/src/main/java/com/example/kevinschweitzer/myapplication/mvp/presenter/CalculatorPresenter.java]: No me estaría andando el operador ternario dentro del método. Parece ser porque setResult esta sobrecargado y no me deja en el mismo método retornar un Int o un String

File: app/src/main/java/com/example/kevinschweitzer/myapplication/mvp/model/CalculatorModel.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    private String number1;
+    private String number2;
+    private Operator operator;
+
+    public void setNewNumber(String str){
+        if(operator == null){
+            if(number1==null)
+                number1=str;
+            else
+                number1+=str;
```
Comment (nacho-globant) [inline app/src/main/java/com/example/kevinschweitzer/myapplication/mvp/model/CalculatorModel.java]: You could minimize this sentence.

number1=(number1==null)?str:number1+str;

File: app/src/main/java/com/example/kevinschweitzer/myapplication/mvp/model/CalculatorModel.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+        }
+        else{
+            if(number2==null)
+                number2=str;
+            else
+                number2+=str;
```
Comment (nacho-globant) [inline app/src/main/java/com/example/kevinschweitzer/myapplication/mvp/model/CalculatorModel.java]: the same

File: app/src/test/java/com/example/kevinschweitzer/myapplication/CalculatorModelTest.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.nullValue;
+import static org.junit.Assert.assertThat;
+
+@RunWith(RobolectricTestRunner.class)
+public class CalculatorModelTest {
+
```
Comment (nacho-globant) [inline app/src/test/java/com/example/kevinschweitzer/myapplication/CalculatorModelTest.java:21]: You should to create another method with the annotation @before and create a unique instante of  CalculatorModel. And use this instance in all test

File: app/src/test/java/com/example/kevinschweitzer/myapplication/CalculatorModelTest.java [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+        assertThat(model.getResult(),is(1));
+
+
+        model.clear();
+
+        /*Substraction with negative result*/
+        model.setNewNumber("1");
+        model.setOperator(new MinusOperator());
+        model.setNewNumber("5");
+        assertThat(model.getResult(),is(-4));
```
Comment (nacho-globant) [inline app/src/test/java/com/example/kevinschweitzer/myapplication/CalculatorModelTest.java:52]: separate in two test methods.

And test the clear method too in another method

File: app/src/test/java/com/example/kevinschweitzer/myapplication/CalculatorModelTest.java [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+
+        model.clear();
+
+        /*Division of an odd number*/
+        model.setNewNumber("5");
+        model.setOperator(new DivisionOperator());
+        model.setNewNumber("2");
+        assertThat(model.getResult(),is(2));
+
+    }
```
Comment (nacho-globant) [inline app/src/test/java/com/example/kevinschweitzer/myapplication/CalculatorModelTest.java]: separate in three methods

File: app/src/test/java/com/example/kevinschweitzer/myapplication/CalculatorPresenterTest.java [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+        presenter.onNumberClicked(NEW_NUMBER);
+
+        verify(model).setNewNumber(NEW_NUMBER);
+        verify(view).concatNumber(NEW_NUMBER);
+        verifyNoMoreInteractions(model,view);
+    }
+
+    @Test
+    public void operatorClickedTest(){
+        Operator operator = Mockito.mock(Operator.class);
```
Comment (nacho-globant) [inline app/src/test/java/com/example/kevinschweitzer/myapplication/CalculatorPresenterTest.java]: Use the annotation @Mock for this

File: app/src/test/java/com/example/kevinschweitzer/myapplication/CalculatorPresenterTest.java [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+        Mockito.when(view.getActivity()).thenReturn(activity);
+
+        presenter.onEqualsClicked();
+
+        verify(view,times(TWO_INVOCATIONS)).getActivity();
+        verify(view).setResult(model.getStringResult());
+        verify(model,times(TWO_INVOCATIONS)).getStringResult();
+
+        verifyNoMoreInteractions(model,view);
+    }
```
Comment (nacho-globant) [inline app/src/test/java/com/example/kevinschweitzer/myapplication/CalculatorPresenterTest.java]: separate in methods for test each case

File: app/src/test/java/com/example/kevinschweitzer/myapplication/CalculatorPresenterTest.java [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+
+        verifyStatic();
+        RxBus.subscribe(eq(activity),any(OperatorObserver.class));
+
+        verifyStatic();
+        RxBus.subscribe(eq(activity),any(EqualsObserver.class));
+
+        verifyStatic();
+        RxBus.subscribe(eq(activity),any(ClearObserver.class));
+
```
Comment (nacho-globant) [inline app/src/test/java/com/example/kevinschweitzer/myapplication/CalculatorPresenterTest.java]: the same

File: app/src/test/java/com/example/kevinschweitzer/myapplication/CalculatorPresenterTest.java [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+        verifyStatic(times(NO_INVOCATIONS));
+        RxBus.clear(any(Activity.class));
+
+        /*With not null activity*/
+        Activity activity = mock(Activity.class);
+        Mockito.when(view.getActivity()).thenReturn(activity);
+        presenter.unregister();
+        verifyStatic();
+        RxBus.clear(activity);
+    }
```
Comment (nacho-globant) [inline app/src/test/java/com/example/kevinschweitzer/myapplication/CalculatorPresenterTest.java:178]: the same

**Review by nacho-globant — APPROVED** [review]

**Pull request merged**